### ISSUE TITLE
Major: Customers seeing "usage limit reached error" for cart rule code incorrectly

### ISSUE THREAD

**Issue by gingerling** [thirtybees/thirtybees#986]
Tldr: suspecting voucher codes only work once, even if you allow 500 uses or several uses per customer. More verification needed. Very serious issue.

I email a code ethical1 to all new customers after their first order, to my knowledge this code was working and had been used, but not very much.

However, I have had two cases recently where a customer called claiming a voucher code (cart rule) is giving them the error "Usage limit reached".

With the first customer (call her Abigail) I did the following:

- Checked the code, ethical1 has not reached the maximum number of usages
- Added more to total available just in case
- Checked the user has not used it before
- Edited the rule so each user can use the code more than once (in case of a cache or cookie issue)
- Asked the uses to recreate the cart in a new private window (same)
- Checked the code wasn't already used on the cart from the back office
- Checked myself and the code DID work for me

None of the fixes worked do in the end I just made Abigail new code, ethical2, which worked. 

However, a few months later I got a second customer (call her Betty) complaining ethical1 wasn't working, same error message.

I did some of the basic checks again, then gave Betty the code, ethical2, I made to resolve Abigails case - but Betty got the error STILL. So I had to create a 3rd code for Betty, ethical3, which worked.

I can't work out exactly what is happening because there is no list of people who used each code in the back office: It's possible that Betty is the next user to try ethical1 since Abigail, or the next to care enough to call when ethical1 didn't work. Indeed ethical1 is automatically emailed to each new customer, but the back office says "total available: 800" right now which, being a round number suggests it's not been used by anyone since I updated the quantity in my attempt to fix Abigails issue.

Betty's issues makes me think that the "Total available" number is perhaps not working and defaulted to 1, as Abigail had used ethical2 (total available: 499). Indeed I tried ethical2 myself and I get the same error - however it is possible that this is caused by me testing the code earlier. 

Most convincingly, I also get the issue with the new third code I made, ethical3 (screenshots)

HOWEVER, this seems to be a new issue that has begun a good while after my last upgrade and moving to TB etc. I am pretty sure this was working before, indeed, I am pretty sure I found the first code worked for me when the first customer complained.

I can't understand what has triggered this issue, I haven't really made changes to my store much in the time, just little tweaks, stuff so minor I couldn't even give you a list.

The issue seems pretty major, if it's widespread.

The error is:

"There is 1 error

    You cannot use this voucher anymore (usage limit reached)"

Screenshots of third code I made.

![Screenshot_20190702-113051_LastPass](https://user-images.githubusercontent.com/2853040/60506211-4a407100-9cbd-11e9-8ff7-490e11d052b6.jpg)

![Screenshot_20190702-113104_Firefox](https://user-images.githubusercontent.com/2853040/60506218-4e6c8e80-9cbd-11e9-8333-288b652da287.jpg)

**Comment by mauroagr** [comment]
Hi,
Try Increase to 10, 20...100 the "Total Available for each user", then test again

**Comment by gingerling** [comment]
Can you clarify, they are set ar 500 and 800, do you want me to try lower numbers?

**Comment by Zengraph** [comment]
Hello, can you update TB on his last version and tell us if the bug still occurs ? I noticed you used v1.0.3 only. Some bugs have already been fixed for vouchers on the last versions.

**Comment by getdatakick** [comment]
> Can you clarify, they are set ar 500 and 800, do you want me to try lower numbers?

Set "Total Available for each user" to higher value than 1. This is different settings than "Total Available"

**Comment by gingerling** [comment]
Yes I did that, same result (that's what I mean by point 4 "Edited the rule so each user can use the code more than once (in case of a cache or cookie issue)")

**Comment by gingerling** [comment]
Did your try verifying? Would you like login details to my store to verify?

**Comment by getdatakick** [comment]
I've tested this functionality and it works fine.

The error message `You cannot use this voucher anymore (usage limit reached)` is displayed only when particular customer already used the voucher, here's [the code](https://github.com/thirtybees/thirtybees/blob/b443f2edd81c65803a4d2704aadeaf79e1622b23/classes/CartRule.php#L1005).

Since you are the only one experiencing this problem, it may be caused by some third party module or override. Can you reproduce the same problem on clean thirtybees installation?

**Comment by gingerling** [comment]
Hi, I'm not really equipped to do that level of testing - is there a way to find the problem using my exiting store?

**Comment by gingerling** [comment]
Tried disabling non-tb plugins using the setting on performance page and it fried my whole site - sysadmin trying to get it back up online

**Comment by gingerling** [comment]
fixed that, and discovered that old vouchers work normally, even if I change the dates/names/etc, it's just new vouchers that don't work right

**Comment by ksimunovic** [comment]
@gingerling Do you still have the same issue? I tried reproducing it with a clean 1.0.3 install but everything works fine.

**Comment by gingerling** [comment]
This error is still intermittently occurring, even on the latest version.
We are also having some successful use of codes.

Becuse it is unreliable, I believe this issue is related to something to do
with the user's browser or cookies, something more complex than a bug that
affects everyone. I don't have any technical evidence for this but it makes
sense based on what I've seen. I think that when the user puts in the code,
something happens, maybe a page refresh, and the user is counted as having
used the code already (even though they just put it in a moment ago).

The reason I am concerned this could be a more widespread bug, is that over
many years of running a website I find that most people don't contact us
about these kind of errors. We often only know there is an issue with, for
example, our payment gateway, when we stop getting orders. People don't
call and say it's broken. If it's an intermittent issue only affecting,
say, 20% of users, it's entirely possible it could go under the radar. It's
made more complicated because there is no way for us to monitor usage of
codes, I can't tell you how often they were used and how often they are
used now, unless I go through every single invoice manually.

**Comment by ksimunovic** [comment]
The problem was in 1.0.3 version which had a bug in code usage check which was fixed with commit [48356c](https://github.com/thirtybees/thirtybees/commit/48356c473dd74277bb6474dca8ea9e372fc6e39e). Problem was fixed by updating to latest version.

**Comment by getdatakick** [comment]
@gingerling can you confirm this is not an issue anymore? Can we closed this one?

**Comment by gingerling** [comment]
Let me check
